### PR TITLE
Add Label and accoutKeyPath to the on chain payment methods

### PR DESCRIPTION
**Pull request by NicolasDorier**
This add a way to specify some fields of the on-chain payment methods such as `Label` and `AccountKeyPath`.

I wanted to use non-primitive (`RootedKeyPath`) types to make the greenfield library easier to use.
However, this caused problem as the default `ObjectModelValidator` would try to visit all the members of the `RootKeyPath` to validate everything. But those were sending back exception.

So I create a new `SkippableObjectModelValidator` which is like the `DefaultObjectModelValidator` of MVC, except that it is possible to register types to skip validations.

I initially wanted to remove all validations, but we rely on it for the UI and there is no way to allow it only for UI and not for API.

### CODE REVIEW

**Comment by Kukks** [comment]
Are you sure about the label? I think it should be a read only property due to the deriv scheme parser converts the xpub input and the original becomes the label

**Comment by NicolasDorier** [comment]
@Kukks this is not readonly. What you refer is the default behavior, but you can set any label in the wallet settings.

**Comment by Kukks** [comment]
> @Kukks this is not readonly. What you refer is the default behavior, but you can set any label in the wallet settings.

TIL

File: BTCPayServer/Hosting/SkippableObjectValidatorProvider.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+using Microsoft.Extensions.Options;
+using NBitcoin;
+
+namespace BTCPayServer.Hosting
+{
+    public class SkippableObjectValidatorProvider : ObjectModelValidator
```
Comment (Kukks) [inline BTCPayServer/Hosting/SkippableObjectValidatorProvider.cs:13]: This is some hardcore stuff

**Review by Kukks — CHANGES_REQUESTED** [review]
Maybe it will simplify the code a lot to simply do the key path parsing and validation inside the controller action instead of the NBitcoin Json converter? It is a lot of hoops to jump through in mvc with this approach

File: BTCPayServer/Hosting/SkippableObjectValidatorProvider.cs (hunk 1 above)
Comment (NicolasDorier) [reply BTCPayServer/Hosting/SkippableObjectValidatorProvider.cs:13]: Well this is possible, but this make the library a pain in the ass to use.
To be honest, I am not really fan of hardcore code like that either. But having to do work for parsing in the controller while we can get NewtonsoftJson do the work for us, making our life easier AND easier to use the client library.

I think for the other actions we'll do (create a hot wallet or querying UTXOs), we'll want to reuse this.

**Review by Kukks — APPROVED** [review]
Ok makes sense then.

File: BTCPayServer/Controllers/GreenField/StoreOnChainPaymentMethodsController.cs [hunk 2]
```diff
@@ -203,6 +204,18 @@ public ActionResult<OnChainPaymentMethodData> GetOnChainPaymentMethod(string cry
                 var strategy = DerivationSchemeSettings.Parse(paymentMethodData.DerivationScheme, network);
                 if (strategy != null)
                     await wallet.TrackAsync(strategy.AccountDerivation);
+                strategy.Label = paymentMethodData.Label;
```
Comment (Kukks) [inline BTCPayServer/Controllers/GreenField/StoreOnChainPaymentMethodsController.cs:207]:
```suggestion
                if(!string.IsNullOrEmpty(paymentMethodData.Label)
                {
                    strategy.Label = paymentMethodData.Label;
                }
```

**Review by Kukks — CHANGES_REQUESTED** [review]
Maybe this last change on the label setting actually?

File: BTCPayServer/Controllers/GreenField/StoreOnChainPaymentMethodsController.cs (hunk 2 above)
Comment (NicolasDorier) [reply BTCPayServer/Controllers/GreenField/StoreOnChainPaymentMethodsController.cs:207]: what I did before, but actually you want to set to empty or null if you want to unset the label.

**Pull request merged**